signer/signer.go

```diff
@@ -0,0 +1,262 @@
+package signer
+
+import (
+	"crypto/x509"
+	"fmt"
+	"time"
+
+	nx509 "github.com/notaryproject/notation-core-go/x509"
+)
+
+type SignatureMediaType string
+
+// SignatureAlgorithm lists supported Signature algorithms.
+type SignatureAlgorithm string
+
+const (
+	RSASSA_PSS_SHA_256 SignatureAlgorithm = "RSASSA_PSS_SHA_256"
+	RSASSA_PSS_SHA_384 SignatureAlgorithm = "RSASSA_PSS_SHA_384"
+	RSASSA_PSS_SHA_512 SignatureAlgorithm = "RSASSA_PSS_SHA_512"
+	ECDSA_SHA_256      SignatureAlgorithm = "ECDSA_SHA_256"
+	ECDSA_SHA_384      SignatureAlgorithm = "ECDSA_SHA_384"
+	ECDSA_SHA_512      SignatureAlgorithm = "ECDSA_SHA_512"
+)
+
+// SignerInfo represents a parsed signature envelope that is agnostic to signature envelope format.
+type SignerInfo struct {
+	Payload            []byte
+	PayloadContentType string
+	SignedAttributes   SignedAttributes
+	UnsignedAttributes UnsignedAttributes
+	SignatureAlgorithm SignatureAlgorithm
+	CertificateChain   []*x509.Certificate
+	Signature          []byte
+	TimestampSignature []byte
+}
+
+// SignedAttributes represents signed metadata in the Signature envelope
+type SignedAttributes struct {
+	SigningTime        time.Time
+	Expiry             time.Time
+	ExtendedAttributes []Attribute
+}
+
+// UnsignedAttributes represents unsigned metadata in the Signature envelope
+type UnsignedAttributes struct {
+	SigningAgent string
+}
+
+// SignRequest is used to generate Signature.
+type SignRequest struct {
+	Payload             []byte
+	PayloadContentType  string
+	CertificateChain    []*x509.Certificate
+	SignatureProvider   SignatureProvider
+	SigningTime         time.Time
+	Expiry              time.Time
+	ExtendedSignedAttrs []Attribute
+	SigningAgent        string
+}
+
+type Attribute struct {
+	Key      string
+	Critical bool
+	Value    interface{}
+}
+
+// SignatureProvider is used to sign bytes generated after creating Signature envelope.
+type SignatureProvider interface {
+	Sign([]byte) ([]byte, error)
+}
+
+type SignatureEnvelope struct {
+	rawSignatureEnvelope []byte
+	internalEnvelope     internalSignatureEnvelope
+}
+
+// Contains a set of common methods that every Signature envelope format must implement.
+type internalSignatureEnvelope interface {
+	// validateIntegrity validates the integrity of given Signature envelope.
+	validateIntegrity() error
+	// getSignerInfo returns the information stored in the Signature envelope and doesn't perform integrity verification.
+	getSignerInfo() (*SignerInfo, error)
+	// signPayload created Signature envelope.
+	signPayload(SignRequest) ([]byte, error)
+}
+
+// Verify performs integrity validation and validates that cert-chain stored in Signature leads to a given set of the trusted root.
+// Returns the certificate that is used for establishing trust i.e. authenticity of the signature, SignerInfo object
+// containing the information about the signature.
+func (s *SignatureEnvelope) Verify(trustedCerts []*x509.Certificate) (*x509.Certificate, *SignerInfo, error) {
```

Comment:
Had an offline discussion to move "authenticity" part of this function into it's own method for flexibility.